Create button guide

gregoryb-at-ibm opened the pull request:
Hi

Is this appropriate for the guides repository?

It's a guide on where to add what code to enable creation of an externally managed item in your EMS.

* Need to determine location and name of file.

Fryguy commented:
> Is this appropriate for the guides repository?

yes, this is the place!

However, I wonder if this might be better in the provider plugins directory as opposed to UI.  @agrare Thoughts on organization?

agrare commented:
> However, I wonder if this might be better in the provider plugins directory

Yeah this could actually be a great next "article" in the writing a new provider series, e.g. https://github.com/ManageIQ/guides/blob/master/providers/writing_a_new_provider.md#next-steps

agrare commented on ui/flow_of_allowing_to_add_models.md:
````diff
@@ -0,0 +1,123 @@
+Make sure you have an application-record in `plugins/manageiq-providers-your-provider/app/models/manageiq/providers/your-provider/your_item.rb`.
+It should extend `app/models/your_item.rb`. 
+
+### Application Record type
+Make sure the database table for the application-record has a "type" column, so that STI can work properly.\
+If it doesn't have such a column, you'll need to add a migration file to `manageiq-schema` and create the column.    
+It will look something like
+```ruby
+class AddTypeToYourItems < ActiveRecord::Migration[6.0]
+  def change
+    add_column :your_items, :type, :string
+  end
+end
+```
````
This is usually handled when adding the new model, not when adding create support.  Maybe better to split up the "How to add a new model" into its own guide which would cover [[lines 22 - 40]](https://github.com/ManageIQ/guides/pull/478/files#diff-7b7fe923f33307d7a91073faa2e44c3b61e1c9c60a8e0db971e67ea7e7a960c6R22-R40)

agrare commented on ui/flow_of_allowing_to_add_models.md:
```diff
@@ -0,0 +1,123 @@
+2. `def self.raw_create_your_item(ext_management_system, options = {})` 
+
+The method `self.raw_create_your_item` performs the actual work of creating.\  
+The options dict will contain what you get from your JS form.\
```
`options dict`

In ruby this is more commonly called a hash, python has dictionaries

agrare commented on ui/flow_of_allowing_to_add_models.md:
```diff
@@ -0,0 +1,123 @@
+
+## Adding support for the new operation 
+1. In `app/models/ext_management_system.rb` add `supports_attribute :supports_create_your_item, :child_model => "YourItem"`
+2. In `manageiq-providers-autosde/app/models/manageiq/providers/your-provider/your_manager.rb`, add `supports :add_your_item`
```
`manageiq-providers-autosde` => best to leave this generic as well like `manageiq-providers-your_provider`

agrare replied:
Actually with ``supports_attribute :supports_create_your_item, :child_model => "YourItem"` the correct way to do this would be:
`manageiq-providers-your_provider/app/models/manageiq/providers/your-provider/your_manager/your_thing.rb` and add `supports :create`

agrare commented on ui/flow_of_allowing_to_add_models.md:
```diff
@@ -0,0 +1,123 @@
+1. In `app/models/ext_management_system.rb` add `supports_attribute :supports_create_your_item, :child_model => "YourItem"`
+2. In `manageiq-providers-autosde/app/models/manageiq/providers/your-provider/your_manager.rb`, add `supports :add_your_item`
+
+## Toolbar Button
```
cc @kavyanekkalapu can you review the UI parts of this?

kavyanekkalapu replied:
@gregoryb-at-ibm UI section Looks good to me.

agrare commented on ui/flow_of_allowing_to_add_models.md:
```diff
@@ -0,0 +1,123 @@
+
+An API controller with a good example of `create_resource` method is `auth_key_pairs_controller.rb`, so you can copy-paste from there and adjust as needed.
+
+If you are adding a new controller, you'll need to add routes to in `manageiq-api/api.yml`. Copy from one of the existing routes and adjust.
```
If you're adding create to the collection you'll also need to edit `api.yml`, think it'd be `:verbs: *gpppd` but that always confuses me :laughing: assume it is `GET, PUT, PATCH, POST, DESTROY` but what if you only have `GET, POST` is it `*gp`?  How do we know that's not `GET, PATCH`? cc @Fryguy

agrare replied:
NOTE you also might need to create a new product feature

Fryguy replied:
yes, `gpppd` is `GET POST PUT PATCH DESTROY`.  The star at the beginning of `*gpppd` is a YAML anchor that points to the definition.  See https://github.com/ManageIQ/manageiq-api/blob/master/config/api.yml#L6-L31  We know it's not `GET PATCH`, because `GET PATCH` is not defined 😉

agrare replied:
:+1: so if you wanted to implement only GET and PATCH you could call them out and just not use the shorthand

Fryguy replied:
If it helps (or for others that aren't aware of YAML anchors)...

```ruby
[1] pry(main)> require "yaml"
=> true
[2] pry(main)> x = {:target => {:a => 1, :b => 2}}
=> {:target=>{:a=>1, :b=>2}}
[3] pry(main)> x[:reference] = x[:target]
=> {:a=>1, :b=>2}
[4] pry(main)> x[:reference].object_id == x[:target].object_id
=> true
[5] pry(main)> puts x.to_yaml
---
:target: &1
  :a: 1
  :b: 2
:reference: *1
=> nil
[6] pry(main)> y = YAML.load(x.to_yaml)
=> {:target=>{:a=>1, :b=>2}, :reference=>{:a=>1, :b=>2}}
[7] pry(main)> y[:reference].object_id == y[:target].object_id
=> true
```

chessbyte replied:
so we created this nomenclature to minimize typing??

chessbyte replied:
anchors are also used by Rails in `config/database.yml`, if I recall correctly

Fryguy replied:
It minimizes _a lot_ of duplication, but yes, it was created to declutter the file.  It also ensures that only known "sets" are chosen.

Fryguy replied:
For reference:

```ruby
[8] pry(main)> YAML.load_file("config/api.yml").deep_dup.to_yaml.lines.count
=> 5457
[9] pry(main)> File.read("config/api.yml").lines.count
=> 5068
```

The first line is fully expanded, the second line is the current, so we're talking 389 lines or 7.5% bigger

gregoryb-at-ibm replied:
@agrare, @Fryguy 
I added instructions about adding `create` to YAML. Could you see that it's correct and complete?

Also, could you help me with instructions for adding an appropriate product feature?

agrare commented on ui/flow_of_allowing_to_add_models.md:
```diff
@@ -0,0 +1,123 @@
+Add the following tests to the spec file. You can copy-paste them from other specs and adjust to your item.
+1. Good path where a POST succeeds in creating a new item.
+2. A create request without a valid ems-id should fail.
+3. A create request for a user without an appropriate role should fail.
```
This would be a good place to mention the product_feature/identifier that you have to authorize the user for in the specs

gregoryb-at-ibm replied:
I've added a section about product features. Could you check it?

agrare commented:
I love this @gregoryb-at-ibm thanks so much for putting the whole flow together!

agrare commented on ui/flow_of_allowing_to_add_models.md:
```diff
@@ -0,0 +1,144 @@
+* `self.raw_create_your_item`
+* `self.create_your_item_queue`
+
+If these aren't present, add these methods to the base record by copying them from some other model in `app/models/`.
```
suggest changing `record` -> `base class` or `base model`

agrare commented on ui/flow_of_allowing_to_add_models.md:
```diff
@@ -0,0 +1,144 @@
+
+### Provider Application Record
+Make sure you have an application-record in `plugins/manageiq-providers-your_provider/app/models/manageiq/providers/your-provider/your_item.rb`.
+It should extend `app/models/your_item.rb` with Rails STI. 
```
Do you want to remove [[this]](https://github.com/ManageIQ/guides/pull/478/files#diff-7b7fe923f33307d7a91073faa2e44c3b61e1c9c60a8e0db971e67ea7e7a960c6R13) from the assumptions if you're also covering it here?  Think it should be in one place or the other but not both

gregoryb-at-ibm replied:
> Looking good @gregoryb-at-ibm just a few minor comments then I'll merge!

I think I fixed all the comments?

Before you merge we should decide on the file's name, and where it should go.
And also I'd like to squash all my commits.

agrare commented on ui/flow_of_allowing_to_add_models.md:
```diff
@@ -0,0 +1,144 @@
+
+## Adding support for the new operation 
+1. In `app/models/ext_management_system.rb` add `supports_attribute :supports_create_your_item, :child_model => "YourItem"`
+2. In `manageiq-providers-your_provider/app/models/manageiq/providers/your-provider/your_item.rb`, add `supports :create`
```
Only need to talk about adding `supports :create` once, [[ref]](https://github.com/ManageIQ/guides/pull/478/files#diff-7b7fe923f33307d7a91073faa2e44c3b61e1c9c60a8e0db971e67ea7e7a960c6R36)

agrare commented on ui/flow_of_allowing_to_add_models.md:
````diff
@@ -0,0 +1,144 @@
+## Product Feature
+Find your item's product feature list in `manageiq/db/fixtures/miq_product_features.yml`. 
+Make sure that it has a section for 'add', as below
+```shell
````
Really minor, but this should be `yaml` not `shell`

agrare commented:
Looking good @gregoryb-at-ibm just a few minor comments then I'll merge!

gregoryb-at-ibm commented:
@agrare 
I think I fixed all the comments?

Before you merge we should decide on the file's name, and where it should go.
Then I'll squash all my commits.

agrare commented:
> Before you merge we should decide on the file's name, and where it should go.

How about `providers/create_operation.md`

gregoryb-at-ibm commented:
> > Before you merge we should decide on the file's name, and where it should go.
> 
> How about `providers/create_operation.md`

OK, moved and squashed.
